src/platforms/hosted/bmda_gpiod.c

```diff
@@ -0,0 +1,264 @@
+		return false;
+	}
+
+	/* This is an unsigned int because that is what gpiod_line_get consumes. */
+	unsigned int gpio_off = (unsigned int)offset_val;
+
+	DEBUG_INFO("gpiochip: %s offset: %u\n", gpio, gpio_off);
+	struct gpiod_line *line = gpiod_line_get(gpio, gpio_off);
+	if (!line) {
+		DEBUG_ERROR("Couldn't get GPIO: %s:%u errno: %d\n", gpio, gpio_off, errno);
+		return false;
+	}
+
+	if (!strcmp("tck", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-tck", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_tck_pin = line;
+	} else if (!strcmp("tms", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-tms", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_tms_pin = line;
+	} else if (!strcmp("tdi", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-tdi", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_tdi_pin = line;
+	} else if (!strcmp("tdo", name)) {
+		if (gpiod_line_request_input_flags(line, "bmda-tdo", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE))
+			goto out_fail_close;
+		bmda_gpiod_tdo_pin = line;
+	} else if (!strcmp("swdio", name)) {
+		if (gpiod_line_request_input_flags(line, "bmda-swdio", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE))
+			goto out_fail_close;
+		bmda_gpiod_swdio_pin = line;
+	} else if (!strcmp("swclk", name)) {
+		if (gpiod_line_request_output_flags(line, "bmda-swclk", GPIOD_LINE_REQUEST_FLAG_BIAS_DISABLE, 0))
+			goto out_fail_close;
+		bmda_gpiod_swclk_pin = line;
+	} else {
+		DEBUG_ERROR("Unrecognised signal name: %s\n", name);
+		gpiod_chip_close(gpiod_line_get_chip(line));
+		return false;
+	}
+
+	DEBUG_INFO("Line consumer: %s\n", gpiod_line_consumer(line));
+
+	return true;
+
+out_fail_close:
+	DEBUG_ERROR("Requesting gpio failed errno: %d", errno);
+	gpiod_chip_close(gpiod_line_get_chip(line));
+	return false;
+}
+
+static bool bmda_gpiod_parse_gpiomap(const char *gpio_map)
+{
+	bool ret = true;
+	char *saveptr;
+	char *gpio_map_copy = strdup(gpio_map);
+	DEBUG_INFO("GPIO mapping: %s\n", gpio_map);
+
+	char *token = strtok_r(gpio_map_copy, ",", &saveptr);
+	while (token) {
+		DEBUG_INFO("GPIO: %s\n", token);
+		char *val = strstr(token, "=");
+		if (!val) {
+			ret = false;
+			break;
+		}
+		*val = '\0';
+		val++;
+		if (!bmda_gpiod_parse_gpio(token, val)) {
+			ret = false;
+			break;
+		}
+
+		token = strtok_r(NULL, ",", &saveptr);
+	}
+
+	free(gpio_map_copy);
+	return ret;
+}
+
+bool bmda_gpiod_init(bmda_cli_options_s *const cl_opts)
+{
+	if (!cl_opts->opt_gpio_map)
+		return false;
+
+	if (!bmda_gpiod_parse_gpiomap(cl_opts->opt_gpio_map))
+		return false;
+
+	if (bmda_gpiod_swclk_pin && bmda_gpiod_swdio_pin) {
+		bmda_gpiod_swd_ok = true;
+	}
+
+	if (bmda_gpiod_tck_pin && bmda_gpiod_tdi_pin && bmda_gpiod_tdo_pin && bmda_gpiod_tms_pin) {
+		bmda_gpiod_jtag_ok = true;
+	}
+
+	return bmda_gpiod_jtag_ok || bmda_gpiod_swd_ok;
+}
+
+bool bmda_gpiod_jtag_init()
```

Comment:
This and the function below get the same comment about needing `void` in the parens to avoid the potentially bad spicy kind of function definition.

Reply:
Done